src/debug.rs

```diff
@@ -0,0 +1,352 @@
+};
+use jsonrpc_core::{BoxFuture, Result};
+use once_cell::sync::OnceCell;
+use std::sync::{Arc, RwLock};
+use vm::{CallTracer, ExecutionResult, HistoryDisabled, TxExecutionMode, Vm};
+use zksync_basic_types::H256;
+use zksync_core::api_server::web3::backend_jsonrpc::{
+    error::into_jsrpc_error, namespaces::debug::DebugNamespaceT,
+};
+use zksync_state::StorageView;
+use zksync_types::{
+    api::{BlockId, BlockNumber, DebugCall, DebugCallType, ResultDebugCall, TracerConfig},
+    l2::L2Tx,
+    transaction_request::CallRequest,
+    PackedEthSignature, Transaction, CONTRACT_DEPLOYER_ADDRESS,
+};
+use zksync_web3_decl::error::Web3Error;
+
+/// Implementation of DebugNamespaceImpl
+pub struct DebugNamespaceImpl<S> {
+    node: Arc<RwLock<InMemoryNodeInner<S>>>,
+}
+
+impl<S> DebugNamespaceImpl<S> {
+    /// Creates a new `Debug` instance with the given `node`.
+    pub fn new(node: Arc<RwLock<InMemoryNodeInner<S>>>) -> Self {
+        Self { node }
+    }
+}
+
+impl<S: Send + Sync + 'static + ForkSource + std::fmt::Debug> DebugNamespaceT
+    for DebugNamespaceImpl<S>
+{
+    fn trace_block_by_number(
+        &self,
+        _block: BlockNumber,
+        _options: Option<TracerConfig>,
+    ) -> BoxFuture<Result<Vec<ResultDebugCall>>> {
+        todo!()
+    }
+
+    fn trace_block_by_hash(
+        &self,
+        _hash: H256,
+        _options: Option<TracerConfig>,
+    ) -> BoxFuture<Result<Vec<ResultDebugCall>>> {
+        todo!()
+    }
+
+    /// Trace execution of a transaction.
+    fn trace_call(
+        &self,
+        request: CallRequest,
+        block: Option<BlockId>,
+        options: Option<TracerConfig>,
+    ) -> BoxFuture<Result<DebugCall>> {
+        let only_top = options.is_some_and(|o| o.tracer_config.only_top_call);
+        let inner = Arc::clone(&self.node);
+        Box::pin(async move {
+            if block.is_some() && !matches!(block, Some(BlockId::Number(BlockNumber::Latest))) {
+                return Err(jsonrpc_core::Error::invalid_params(
+                    "tracing only supported at `latest` block",
+                ));
+            }
+
+            let inner = inner
+                .read()
+                .map_err(|_| into_jsrpc_error(Web3Error::InternalError))?;
+
+            let mut l2_tx = match L2Tx::from_request(request.into(), MAX_TX_SIZE) {
+                Ok(tx) => tx,
+                Err(e) => {
+                    let error = Web3Error::SerializationError(e);
+                    return Err(into_jsrpc_error(error));
+                }
+            };
+
+            let execution_mode = TxExecutionMode::EthCall;
+            let storage = StorageView::new(&inner.fork_storage).to_rc_ptr();
+
+            let bootloader_code = inner.system_contracts.contacts_for_l2_call();
+
+            // init vm
+            let (l1_batch_env, _block_context) = inner.create_l1_batch_env(storage.clone());
+            let system_env = inner.create_system_env(bootloader_code.clone(), execution_mode);
```

Comment:
We should consider updating the `gas_limit` per the comments from the `zksync_core` implementation: 
https://github.com/matter-labs/zksync-era/blob/main/core/lib/zksync_core/src/api_server/execution_sandbox/execute.rs#L93-L96

``` rust
    // Protection against infinite-loop eth_calls and alike:
    // limiting the amount of gas the call can use.
    // We can't use BLOCK_ERGS_LIMIT here since the VM itself has some overhead.
    l2_tx.common_data.fee.gas_limit = ETH_CALL_GAS_LIMIT.into();
```